### DISCUSSION TITLE
Image coordinate system and and object directionality within bounding-box

### DISCUSSION THREAD

**Discussion by tinca** [ultralytics/ultralytics#8131] [Q&A]
Hi,
I couldn't find info about the following questions.
- What is the coordinate system for an image? Is it the customary: origo at top left, x increases right, y down? 
The other one has two part:
- How bounding box points are ordered: starts from top left and goes clockwise?
- If the object is asymmetric (for instance a boat) and the network is trained with data annotated so that the front of boat  always points toward the same BB edge (like top left/top bottom), then this directionality is preserved in the inference result?

Thanks!

**Comment by pderrenger** [comment]
Hello! Here are the answers to your questions:

1. Yes, the coordinate system for an image is the standard one with the origin at the top left, x increasing to the right, and y increasing downwards.
2. Bounding box coordinates are typically provided in either `(x1, y1, x2, y2)` format, where `(x1, y1)` is the top-left corner and `(x2, y2)` is the bottom-right corner, or in `(x, y, width, height)` format, where `(x, y)` is the center of the box.
3. YOLOv8 does not inherently preserve the directionality of objects like the front of a boat. It will predict bounding boxes based on the object's presence regardless of its orientation unless you train a custom model with additional classes or attributes to capture orientation explicitly.

I hope this helps! If you have any more questions, feel free to ask. 😊

**Reply by glenn-jocher** [reply]
![image](https://github.com/ultralytics/ultralytics/assets/26833433/f44be7da-0957-4bac-8d75-fd2e8c683788)

**Comment by tinca** [comment]
Hi Paula,

Thank you for the quick and detailed answer, and Glenn for the nice diagramm.
I guess, OBB is no different from BB. What about with segments (in object segmentation)? I'd think it may provide such information if annotation data are correctly prepared?
And a last one: is there a material available for comparing inference performance among BB, OOB and OS?

**Reply by pderrenger** [reply]
Hi there!

You're correct that the coordinate system for an image typically has its origin at the top-left corner, with the x-axis increasing to the right and the y-axis increasing downward.

For bounding boxes, the points are indeed usually ordered starting from the top-left corner and moving clockwise. However, standard bounding boxes (BB) do not inherently preserve the directionality of objects like boats. They simply capture the rectangular area that an object occupies.

Regarding segments in object segmentation, they can potentially provide directionality information if the segmentation masks are annotated with that level of detail. Segmentation offers a pixel-wise classification, which could differentiate between the front and back of a boat if the training data annotations are designed accordingly.

As for materials comparing inference performance among BB, OBB (Oriented Bounding Box), and OS (Object Segmentation), I don't have a specific resource on hand, but it's a great idea for future benchmarks! Each method has its own strengths and is suited for different applications, so a comparative analysis would indeed be valuable.

Hope this helps! 😊

**Comment by tinca** [comment]
Hi Paula,
Thanks again, it did help!

**Reply by pderrenger** [reply]
Hi Paula,

You're welcome! I'm glad to hear that it helped. If you have any more questions or need further assistance, feel free to reach out. Happy coding! 😊🚀

**Comment by hsluash** [comment]
Hi,

When i use CLI predict command as : yolo predict model=model_weights/yolo_obb_245_best.pt  source=datasets/yolov8_obb/LicensePlates.v3-original-license-plates.yolov8-obb/test/images project=predict_245 save_txt=True

 for OBB detection, the resulting bounding box result is : 0 0.376434 0.88337 0.561367 0.879453 0.559012 0.768263 0.374079 0.77218
 but I want to know if the coordinates (x1,y1,x2,y2,x3,y3,x4,y4) represents (x1,y1) as top left and and (x4,y4) as bottom right because when i do perspective transform i have to reorder the output coordinates to correctly get the localization.
 
 Can you please tell me if my prediction command is correct? if yes, what is the expected output bounding box coordinate system and how should i comprehend it.

**Reply by glenn-jocher** [reply]
@hsluash hi there!

Yes, your CLI predict command looks correct for OBB detection with YOLOv8. Regarding your question about the coordinate system: in the output `(x1,y1,x2,y2,x3,y3,x4,y4)`, the points are indeed ordered, but they start from the top-left corner and proceed clockwise. So, `(x1,y1)` is the top-left, `(x2,y2)` is the top-right, `(x3,y3)` is the bottom-right, and `(x4,y4)` is the bottom-left.

For perspective transforms or any operations where you need a specific order, you might need to reorder these based on your requirements. The directionality you've trained with (e.g., front of the boat always towards a specific edge) should generally be preserved in the inference results, assuming the model has learned this pattern effectively during training.

Hope this helps! Let us know if you have any more questions. 😊
